### PR TITLE
Implementation for atomkraft reactor

### LINKED ISSUE #22: Implement ADR 04
Implement the `atomkraft reactor` command, as described in [ADR04](https://github.com/informalsystems/atomkraft/blob/ivan/adr04Reactor/docs/src/adr/04adr-reactor-stub.md)

### PR DESCRIPTION
**Pull request by ivan-gavran**

### CODE REVIEW

File: atomkraft/cli/__init__.py [hunk 1]
```diff
@@ -2,9 +2,11 @@
 
 import pytest
 import typer
+import os
 from copier import run_auto
 
 from .. import chain
+from .. import Reactor
```
Comment (rnbguy) [inline atomkraft/cli/__init__.py]: Python convention for naming modules is [lowercase names with underscores](https://peps.python.org/pep-0008/#package-and-module-names).

Please rename `atomkraft/Reactor.py` to `atomkraft/reactor.py` and update the imports accordingly.

File: atomkraft/Reactor.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
+from os import PathLike
+from typing import List
+
+
+def generate_reactor(
+    actions_list: List[str], variables_list: List[str], stub_file_path: PathLike = None
```
Comment (rnbguy) [inline atomkraft/Reactor.py]: `stub_file_path: PathLike = None` should be `stub_file_path: Optional[PathLike] = None` or `stub_file_path: PathLike | None = None`.

**Review by rnbguy — COMMENTED** [review]
Added some comments.

Otherwise, looks good to me :+1:

File: atomkraft/reactor_room/reactor_room.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        constants.ATOMKRAFT_INTERNAL_CONFIG,
+    )
+    with open(internal_config_file_path) as config_f:
+        config_data = tomlkit.load(config_f)
+        return config_data[constants.REACTOR_CONFIG_KEY]
+
+
+def generate_reactor(
+    actions_list: List[str], variables_list: List[str], stub_file_path: PathLike = None
+) -> PathLike:
```
Comment (andrey-kuprianov) [inline atomkraft/reactor_room/reactor_room.py]: I believe the API here needs to be changed from:

```python
def check_reactor(trace: PathLike, keypath="action", reactor=None) -> bool:

def get_reactor() -> PathLike

def generate_reactor(
    actions_list: List[str], variables_list: List[str], stub_file_path: PathLike = None
) -> PathLike
```

to

```python
def check_reactor(trace: PathLike, reactor: PathLike) -> bool:

def get_reactor() -> PathLike

def generate_reactor(
    actions_list: List[str], variables_list: List[str], keypath="action", stub_file_path: PathLike = None
) -> PathLike
```

The point is that the `keypath` argument should be considered as part of the reactor. The user, when they call `test trace`, which under the hood calls `get_reactor` and `check_reactor` should not provide the `keypath` argument each time. Thus, it should be moved to the `generate_reactor` function.

**Review by andrey-kuprianov — CHANGES_REQUESTED** [review]

**Comment by andrey-kuprianov** [comment]
also, how about renaming `reactor_room` to just `reactor`? ;) I know, it's not too important, but `reactor_room` is too long for my taste, and the `_room` suffix doesn't _really_ convey any additional meaning...

File: atomkraft/reactor_room/reactor_room.py (hunk 3 above)
Comment (ivan-gavran) [reply atomkraft/reactor_room/reactor_room.py]: I  agree for the workflow of `test trace`. But what about when the user wrote his own reactor and wants to check it against a trace?  

What I could do is to give this signature
`def check_reactor(trace: PathLike, keypath=None, reactor=None) -> bool:`, and if `keypath` is `None`, then to look into the config file (where I would store the `keypath` upon creating the reactor).

Comment (ivan-gavran) [reply atomkraft/reactor_room/reactor_room.py]: resolved synchronously: `keypath` should be a part of the generated file (I misunderstood that part initially).

**Comment by andrey-kuprianov** [comment]
merged `dev` here again to facilitate reviewing

**Review by andrey-kuprianov — APPROVED** [review]
Great work, Ivan! I know this checking of matching actions between trace & reactor did cost some effort, but it's worth to have this check.

Looks good to go from my side. I will be away now for a couple of hours, so feel free either to merge as is, or to first do the small modification wrt. `keypath`, as you prefer.

File: atomkraft/cli/__init__.py [hunk 4]
```diff
@@ -25,3 +28,27 @@ def init(path: Path):
...
+        ..., help="A list of actions for which to create reactor stubs."
+    ),
+    reactor_stub_file: str = typer.Option(
+        os.path.abspath("reactors/reactor.py"),
+        help="A path where to create the reactors file.",
+    ),
+):
+    actions = [act.strip() for act in actions_list.split(",")]
+    variables = [var.strip() for var in variables_list.split(",")]
+    reactor.generate_reactor(actions, variables, reactor_stub_file)
```
Comment (andrey-kuprianov) [inline atomkraft/cli/__init__.py]: How about simplifying this a bit?

```suggestion
    actions: str = typer.Argument(
        ..., help="trace actions for which to create reactor stub functions.",
        show_default=False
    ),
    variables: str = typer.Argument(
        ..., help="state variables to use as parameters for reactor stub actions",
        show_default=False
    ),
    path: typer.FileTextWrite = typer.Option(
        os.path.abspath("reactors/reactor.py"),
        help="path where to create the reactor stub",
    ),
):
    '''
    Generate reactor stub used to interpret test traces
    '''
    actions = [act.strip() for act in actions.split(",")]
    variables = [var.strip() for var in variables.split(",")]
    reactor.generate_reactor(actions, variables, path.name)
```

**Review by andrey-kuprianov — APPROVED** [review]
It looks great, Ivan! The right point to go to a vacation:)

**Pull request merged**